feat(cli): add golden tests for errors (#11588)

elasticspoon opened the pull request:
Fixes #11588

Creates golden files from `coder/cli/errors.go`.
Adds a unit test to test against golden files.
Adds a make file command to regenerate golden files. 
Abstracts test against golden files.

~~Most of the error printing is handled by `cli.prettyErrorFormatter` and `prettyErrorFormatter.format(err)`, neither is exported. I wanted to add tests in `root_internal_test.go` but that resulted in a circular dependency since it needed `clitest`.~~

~~I solved this issue by creating an `export_cli_test.go` file in which I exported those values. Since it is an `_test` file its exports will limited to `cli_test` package files.~~

Exported the formatted for testing purposes.

elasticspoon committed:
feat(cli): add golden tests for errors (#11588)

Creates golden files from `coder/cli/errors.go`.
Adds a unit test to test against golden files.
Adds a make file command to regenerate golden files.
Abstracts test against golden files.

fix: prevent data race

refactor: make formatter public
Parse error command in a less brittle way

mafredri commented:
Thanks for the PR @elasticspoon. There seems to be a data race, caught by CI, would you mind taking a look at that?

elasticspoon commented:
Fixed

Emyrk reviewed (commented):
This all looks good :+1:. Was just talking about getting this in, thanks!

Let's just export the original `prettyErrorFormatter` and the `new`. If we can import that directly, it'll make code navigation easier.

You will have to rebase on `main` to get the latest error change update as well.

Ping me with any updates, happy to fast track this in :+1: 

@elasticspoon

elasticspoon commented:
@Emyrk addressed the issues. And exporting the formatted did clean it up a decent amount. Only other concern I have is:

```go
if tt.Name == "coder exp example-error multi-multi-error" {
  inv.Stdin = os.Stdin
}
```
I have that line in my test because the multi-multi error does:
```go
{
Use:   "multi-multi-error",
Short: "This is a multi error inside a multi error",
Handler: func(inv *serpent.Invocation) error {
  // Closing the stdin file descriptor will cause the next close
  // to fail. This is joined to the returned Command error.
  if f, ok := inv.Stdin.(*os.File); ok {
  _ = f.Close()
  }
  
  return errors.Join(
    xerrors.Errorf("first error: %w", errorWithStackTrace()),
    xerrors.Errorf("second error: %w", errorWithStackTrace()),
  )
},
```
I would like to change the example error to something like:
```go
return errors.Join(
  xerrors.Errorf("parent error: %w", errorWithStackTrace()),
  errors.Join(
	  xerrors.Errorf("child first error: %w", errorWithStackTrace()),
	  xerrors.Errorf("child second error: %w", errorWithStackTrace()),
  ),
)
```

but I am unsure if this changes the meaning of what is being tested. 

Is the example test only for the rendering of error or should it also be testing that the errors get joined together correctly?

Emyrk commented:
@elasticspoon Change the test to what you suggested. 

```golang
> ```go
> return errors.Join(
>   xerrors.Errorf("parent error: %w", errorWithStackTrace()),
>   errors.Join(
> 	  xerrors.Errorf("child first error: %w", errorWithStackTrace()),
> 	  xerrors.Errorf("child second error: %w", errorWithStackTrace()),
>   ),
> )
> ```
````

> but I am unsure if this changes the meaning of what is being tested.

It's all fine imo. The `example-error` is purely for testing error formatting. If there is a reason to bring back the `stdin.Close()`, we can bring it back later. 

Make that change, drop the condition, and I'll green check this :+1:

elasticspoon committed:
refactor: remove test conditional

Change multi-multi-error do not produce an error
using a second close of os.Stdin

elasticspoon commented:
@Emyrk done

Emyrk commented:
Merged! Thanks @elasticspoon !
